### ISSUE TITLE
Make configuration preparation independent of runtime scientific imports

### ISSUE THREAD

**Issue by maindlt** [dynamics-of-stellar-systems/tnt#56]
## Context

Configuration preparation is intended to load, resolve, and validate configuration data without constructing scientific runtime objects. Its current import path nevertheless initializes much of TNT's runtime stack. In particular:

- importing any `tnt.*` submodule first executes `tnt/__init__.py`;
- configuration validation consumes potential, kinematics, and parameter-generator registration metadata;
- importing those package/submodule paths currently loads concrete runtime classes and, transitively, `galax`, Equinox, JAX, MGE, and related runtime code.

PR #55 made registered classes the authoritative source for configured type dispatch and validation metadata. That is the right single-source design, but it also makes the existing configuration/runtime import coupling visible. The coupling is not a blocker for PR #55 and this issue is non-urgent.

## Goal

Allow configuration-only workflows to import and prepare TNT configuration without importing `galax` or initializing runtime potential classes. Preserve full type-name, parameter-schema, and required-settings validation from authoritative declarations.

Potential benefits include:

- faster, lighter configuration inspection and validation;
- fewer import-time side effects;
- a clearer boundary between plain configuration/schema handling and runtime-object construction;
- fewer circular-import risks as TNT grows;
- configuration tooling that does not need to initialize the complete scientific runtime.

## Important design constraint

Making only `tnt/__init__.py` lazy is not sufficient. Python executes `tnt.potential.__init__` and `tnt.kinematics.__init__` before importing their registry submodules, and those initializers currently import concrete classes so decorators populate class registries.

The solution must therefore preserve one authoritative declaration per configured type without reintroducing parallel hand-maintained lists. Possible directions include:

- lightweight registry entries containing schema metadata plus a lazy runtime-class loader;
- separating static configuration metadata from runtime-class loading while binding both through one registration declaration;
- lazy package exports (`__getattr__` or equivalent) combined with registry initialization that does not require concrete scientific classes;
- another design that provides the same single-source guarantee.

Do not solve this by copying type names or schemas back into configuration validation.

## Scope decision

Decide explicitly whether the target is:

1. **galax-free configuration preparation**: `galax` and concrete potential runtime classes are not imported, while some JAX-related dependencies may still load; or
2. **scientific-runtime-free configuration preparation**: configuration-only imports also avoid JAX and Equinox initialization.

The first target is narrower and likely more practical. The second has greater isolation value but interacts with `tnt.numerics`, `unxt`, unit validation, and TNT's process-wide JAX precision policy.

## Suggested acceptance criteria

- A fresh-process import/configuration test demonstrates the agreed boundary, for example checking `sys.modules` after importing `tnt.configuration` and preparing a representative configuration.
- Configuration preparation retains current type, registry-reference, parameter-schema, unit, and generator-setting validation.
- Runtime imports still populate every class registry and construct the same runtime objects.
- No parallel type-name or schema lists are introduced.
- Both configuration-first and runtime-first import orders work without circular imports.
- The chosen boundary and extension workflow are documented in `aidocs/KNOWLEDGE.md`.

## Relationships

- Follow-up to issue #46 / PR #55, which establishes explicit class registration and removes parallel dispatch metadata.
- Related to issue #15 only insofar as both concern configuration architecture; this issue does not change merge semantics or type-conditional defaults.

Assisted by Codex (OpenAI).

**Comment by prashjet** [comment]
**Assessment after mapping the actual import graph**

**The current state is cleaner than this issue assumes.** An AST scan of the whole `tnt/` package for non-`TYPE_CHECKING`, non-module-level imports returns exactly one:

```
tnt/units.py:170  from tnt.potential.registry import parameter_schema_is_known, raw_parameter_dimensions
```

No other lazy `tnt.*` imports, no lazy third-party imports, and no runtime circular imports anywhere. So the concrete debt here is a single lazy import plus its apologetic comment — not a pervasive coupling problem.

**Proposed reframing of the goal.** "Configuration-only workflows that don't import galax" isn't a compelling end in itself — config preparation is always a prelude to a run, galax/JAX/equinox are hard dependencies that are always installed, and `import tnt` already commits to a JAX precision policy before configuration is touched. What *is* worth doing: remove the lazy import, keep the graph acyclic in both import orders, and don't reintroduce parallel lists. I'd narrow this issue to that.

**The narrow fix.** `tnt/units.py` currently carries two unrelated jobs:

1. unit-system + declared-quantity primitives (`UnitSystems`, `build_unit_systems`, `validate_dimension`, `declared_quantity`, `resolve_cosmological_parameters`, …) — a true leaf, `unxt` + `tnt.validation` only, imported low in the graph;
2. `validate_configuration_quantities` + `_potential_parameter_dimensions` + `_validate_field` + `_validate_parameter_units` — whole-config-dict validation, sibling to `_validate_potential` / `_validate_units` in `tnt/configuration/validation.py`, with that file as its only caller.

Job 2 is the *only* reason for the lazy import: it needs `tnt.potential.registry`, and a module-level import there closes `tnt.units → tnt.potential → tnt.mge → tnt.units` — but only because the Job-1 primitives sit below `tnt.mge`. `tnt/configuration/validation.py` already imports `tnt.potential.registry` at module level with no cycle.

So: **move `validate_configuration_quantities` and its private helpers (~90 LOC) from `tnt.units` into `tnt.configuration.validation`.** Then the lazy import becomes a plain one, `tnt.units` drops to a leaf with zero `tnt.potential` coupling, and the check-ordering rationale (currently split across the two files) becomes local. One wrinkle: `_validate_parameter_units` uses the private `_validated_declared_unit`, which must stay in `units.py` for `build_unit_systems` — promote it to a public `validate_declared_unit(value, dimension, path)` primitive. Tests: move the four quantity-validation cases from `test_units.py` to `test_configuration.py`.

**Explicitly out of scope: galax-free config preparation.** Achieving that (config validation reads type schema metadata without importing the runtime classes) requires splitting each configured type's static metadata from its runtime class + converters — which re-fragments the single-source design PR #55 just established, or needs a lazy-class-binding mechanism to avoid that. The payoff is ~1–2 s off `import tnt.configuration` and a firmer boundary, against no identified workflow that needs it. Recommend deferring until a concrete config-tooling need appears.

**Adjacent, optional:** `tnt/validation.py` is a clean primitives leaf except it pulls `jax.numpy` / `numpy` / `astropy` for four data-file check helpers (`_read_bin_ids`, `_validated_bin_ids`, `_finite`, `_positive_finite`) used by `tnt.kinematics` / `tnt.populations`. Moving those to a `tnt/kinematics/_data_checks.py` would make `tnt.validation` stdlib-only. Not a cycle issue — purely tidiness, mention only for completeness.

**Revised acceptance criteria**

- `validate_configuration_quantities` (+ helpers) relocated to `tnt.configuration.validation`; the `tnt/units.py:170` lazy import removed.
- `tnt.units` imports nothing from `tnt.potential` (or any runtime-family module) in either direction.
- Both configuration-first and runtime-first import orders remain acyclic (fresh-process test).
- No parallel type-name or schema lists introduced.
- `aidocs/KNOWLEDGE.md` updated to reflect `tnt.units` as a leaf and record that galax-free config prep was considered and deferred.

**Comment by maindlt** [comment]
Follow-up assessment after checking the current import graph:

The comment above identifies a genuinely small and useful cleanup, but it does so by narrowing the original goal. Moving `validate_configuration_quantities` and its private helpers from `tnt.units` into `tnt.configuration.validation` would remove the one function-local potential-registry import, keep `tnt.units` as a cleaner low-level leaf, and localize whole-configuration validation. That is a focused, low-risk refactor and can be implemented independently without a linked issue.

It would not, however, make configuration preparation independent of the scientific runtime. A fresh Linux process importing `tnt.configuration` still loads JAX, Equinox, `galax`, `tnt.mge`, the concrete potential stack, and concrete kinematics classes. The AST result counts function-local imports; it does not capture the weight of ordinary module-level package initialization. In particular, importing registry submodules first executes the `tnt.potential` and `tnt.kinematics` package initializers, which populate registries by importing concrete classes.

Therefore the narrow units cleanup is much smaller than initially thought, while the issue as titled remains a larger registry/package-initialization redesign. We are deliberately deferring that broader galax-free or scientific-runtime-free configuration boundary until a concrete configuration-tooling need justifies it. The wontfix label records that decision; it should not be interpreted as saying the narrow cleanup is undesirable.

Assisted by Codex (OpenAI).

**Comment by maindlt** [comment]
PR #64 is now merged. Whole-configuration quantity validation lives in `tnt.configuration.validation`; reusable unit primitives remain in `tnt.units`, which no longer imports the potential registry. Fresh-process tests cover both import orders.

This completes the narrow cleanup discussed above. Configuration preparation still imports scientific runtime classes through registry initialization, so the broader goal of this issue remains deferred until a concrete configuration-tooling need justifies it.

Assisted by Codex (OpenAI).
